cmd/rpc-common.go

```diff
@@ -33,19 +34,73 @@ func isRequestTimeAllowed(requestTime time.Time) bool {
 		utcNow.Sub(requestTime) > rpcSkewTimeAllowed)
 }
 
+// semVersion - RPC semantic versioning.
+type semVersion struct {
+	Major uint64
+	Minor uint64
+	Patch uint64
+}
+
+// semver comparator implementation based on the semver 2.0.0 https://semver.org/.
+func (v semVersion) Compare(o semVersion) int {
+	if v.Major != o.Major {
+		if v.Major > o.Major {
+			return 1
+		}
+		return -1
+	}
+	if v.Minor != o.Minor {
+		if v.Minor > o.Minor {
+			return 1
+		}
+		return -1
+	}
+	if v.Patch != o.Patch {
+		if v.Patch > o.Patch {
+			return 1
+		}
+		return -1
+	}
+	return 0
+}
+
+func (v semVersion) String() string {
+	b := make([]byte, 0, 5)
+	b = strconv.AppendUint(b, v.Major, 10)
+	b = append(b, '.')
+	b = strconv.AppendUint(b, v.Minor, 10)
+	b = append(b, '.')
+	b = strconv.AppendUint(b, v.Patch, 10)
+	return string(b)
+}
+
 // AuthRPCArgs represents minimum required arguments to make any authenticated RPC call.
 type AuthRPCArgs struct {
 	// Authentication token to be verified by the server for every RPC call.
 	AuthToken string
+	Version   semVersion
 }
 
 // SetAuthToken - sets the token to the supplied value.
 func (args *AuthRPCArgs) SetAuthToken(authToken string) {
 	args.AuthToken = authToken
 }
 
+// SetRPCVersion - sets the rpc version to the supplied value.
+func (args *AuthRPCArgs) SetRPCVersion(version semVersion) {
+	args.Version = version
+}
+
 // IsAuthenticated - validated whether this auth RPC args are already authenticated or not.
 func (args AuthRPCArgs) IsAuthenticated() error {
+	// checks if rpc Version is not equal to current server rpc version.
+	// this is fine for now, but in future when we add backward compatible
+	// APIs we need to make sure to allow lesser versioned clients to
+	// talk over RPC, until then we are fine with this check.
+	if args.Version.Compare(globalRPCVersion) != 0 {
```

Comment:
It is better if we do this check after token authentication to hide server's rpc version from malicious users.

Reply:
Ok